File: interpreter/monkey/token/token.go
package token

type TokenType string

type Token struct {
    Type TokenType  // a type to differentiate the type of the token
    Literal string  // the literal value of the token. using a string is verstatile but won't be as performant as using int or char.
}

// Define the token type constants
const (
    ILLEGAL = "ILLEGAL"
    EOF = "EOF"

    // Identifiers and literals
    IDENT = "IDENT"
    INT = "INT"

    // Operators
    ASSIGN = "="
    PLUS = "+"
    MINUS = "-"
    BANG = "!"
    ASTERISK = "*"
    SLASH = "/"

    LT = "<"
    GT = ">"

    // Delimiters
    COMMA = ","
    SEMICOLON = ";"

    LPAREN = "("
    RPAREN = ")"
    LBRACE = "{"
    RBRACE = "}"

    // Keywords
    FUNCTION = "FUNCTION"
    LET = "LET"
    TRUE = "TRUE"
    FALSE = "FALSE"
    IF = "IF"
    ELSE = "ELSE"
    RETURN = "RETURN"
)

var keywords = map[string]TokenType{
    "fn": FUNCTION,
    "let" : LET,
    "true" : TRUE,
    "false" : FALSE,
    "if" : IF,
    "else" : ELSE,
    "return" : RETURN,
}

func LookupIdent(input string) TokenType {
    if tok, ok := keywords[input]; ok { 
        return tok 
    }
    return IDENT 
}

File: interpreter/c_impl/token.c
#include <stdlib.h>
#include <stdio.h>

typedef char* TOKEN_TYPE;

#define ILLEGAL "ILLEGAL"
#define INT "INT"
#define ASSIGN "="
#define PLUS "+"
#define COMMA ","
#define SEMICOLON = ";"
#define LPAREN "("
#define RPAREN ")"
#define LBRACE "{"
#define RBRACE "}"
#define FUNCTION "FUNCTION"
#define LET "LET"

struct Lexer{
    char* input;
    int position;
    int readPosition;
    char ch;
};

void read_char(struct Lexer* l) {
    if (l->readPosition >= l->input[sizeof(l->input)]) {
        l->ch = '\0';
    }
    else {
        l->ch = l->readPosition;
    }
    printf("%c", l->input[l->readPosition]);
    l->position = l->readPosition;
    l->readPosition += 1;
}

struct Lexer* new(char* input) {
    struct Lexer* l = malloc(sizeof(*l)); 
    l->input = input;
    read_char(l);
    return l;
}

int main() {
    char* input = "+";
    new(input);
    return 0;
}
